WordPressAuthenticator/Analytics/SignInTracker.swift

```diff
@@ -0,0 +1,341 @@
+    }
+    
+    /// Track a failure.
+    ///
+    func track(failure: String) {
+        track(event(failure: failure))
+    }
+    
+    // MARK: - Event Construction & Context Updating
+    
+    /// Creates an event for a step.  Updates the state machine.
+    ///
+    /// - Parameters:
+    ///     - step: the step we're tracking..
+    ///     - flow: the flow that the step belongs to.
+    ///
+    /// - Returns: an analytics event representing the step.
+    ///
+    private func event(step: Step, flow: Flow) -> AnalyticsEvent {
+        let event = AnalyticsEvent(
+            name: EventType.step.rawValue,
+            properties: properties(step: step, flow: flow))
+        
+        saveLastProperties(step: step, flow: flow)
+        
+        return event
+    }
+
+    /// Creates an event for a failure.  Loads the properties from the state machine.
+    ///
+    /// - Parameters:
+    ///     - failure: the error message we want to track.
+    ///
+    /// - Returns: an analytics event representing the failure.
+    ///
+    private func event(failure: String) -> AnalyticsEvent {
+        var properties = lastProperties()
+        properties[Property.failure.rawValue] = failure
+        
+        return AnalyticsEvent(
+            name: EventType.failure.rawValue,
+            properties: properties)
+    }
+    
+    /// Creates an event for a click interaction.  Loads the properties from the state machine.
+    ///
+    /// - Parameters:
+    ///     - click: the target of the click interaction.
+    ///
+    /// - Returns: an analytics event representing the click interaction.
+    ///
+    private func event(click: ClickTarget) -> AnalyticsEvent {
+        var properties = lastProperties()
+        properties[Property.click.rawValue] = click.rawValue
+
+        return AnalyticsEvent(
+            name: EventType.interaction.rawValue,
+            properties: properties)
+    }
+    
+    // MARK: - Source
+    
+    func set(source: Source) {
+        context.lastSource = source
+    }
+    
+    // MARK: - Properties
+    
+    private func properties(step: Step, flow: Flow) -> [String: String] {
+        return properties(step: step, flow: flow, source: context.lastSource)
+    }
+    
+    private func properties(step: Step, flow: Flow, source: Source) -> [String: String] {
+        return [
+            Property.flow.rawValue: flow.rawValue,
+            Property.source.rawValue: source.rawValue,
+            Property.step.rawValue: step.rawValue,
+        ]
+    }
+    
+    // MARK: - Properties: state machine
+    
+    /// Retrieve the last step, flow and source stored in the state machine.
+    ///
+    private func lastProperties() -> [String: String] {
+        return properties(step: context.lastStep, flow: context.lastFlow, source: context.lastSource)
+    }
+    
+    /// Save the step and flow in the state machine.  The source can only be changed directly using `set(source:)`.
```

Comment:
>The source can only be changed directly using `set(source:)`.

I do appreciate good guard rails 😀

Reply:
I'm still not 100% sure it's the best approach, but that was the intention!  If it needs to change we'll be able to re-evaluate in the actual implementation of the tracking.